blocks/main_loop.py

```diff
@@ -73,13 +77,14 @@ def run(self):
                 self._run_extensions('before_training')
                 self.algorithm.initialize()
                 self.status._training_started = True
+            else:
+                self._run_extensions('on_resumption')
             while self._run_epoch():
                 pass
-        except KeyboardInterrupt:
-            self._run_extensions('on_interrupt')
         except TrainingFinish:
             self.log.current_row.training_finished = True
         finally:
+            signal.signal(signal.SIGINT, default_action)
```

Comment:
Finally have a bit of time to read the main loop code thoroughly. Just wondering what's happening here: `_handle_keyboard_interrupt` runs the `on_interrupt` extensions and makes a record in the log. Training then finishes up the rest of the batch before `_check_finish_training` finds the log entry, and raises `TrainingFinish`. This causes another log entry to be added (`training_finished`)... But then `default_action` gets called, killing the main loop before `after_training` extensions get called.

Where would you save the state of the model? You don't want to do it in `on_interrupt`, because you could be in the middle of anything. But Python dies before the `after_training` extensions get run.

P.S. Instead of storing `default_action`, I think you can just use `signal.SIG_DFL` handler.